src/librustc_resolve/build_reduced_graph.rs

```diff
@@ -99,154 +99,259 @@ impl<'a> Resolver<'a> {
         let path = &use_tree.prefix;
 
-        let mut module_path: Vec<_> = prefix.segments.iter()
+        let module_path: Vec<_> = prefix.segments.iter()
             .chain(path.segments.iter())
             .map(|seg| seg.ident)
             .collect();
 
-        match use_tree.kind {
-            ast::UseTreeKind::Simple(rename, ..) => {
-                let mut ident = use_tree.ident();
-                let mut source = module_path.pop().unwrap();
-                let mut type_ns_only = false;
-
-                if nested {
-                    // Correctly handle `self`
-                    if source.name == keywords::SelfValue.name() {
-                        type_ns_only = true;
-
-                        let last_segment = *module_path.last().unwrap();
-                        if last_segment.name == keywords::CrateRoot.name() {
-                            resolve_error(
-                                self,
-                                use_tree.span,
-                                ResolutionError::
-                                SelfImportOnlyInImportListWithNonEmptyPrefix
-                            );
-                            return;
-                        }
-
-                        // Replace `use foo::self;` with `use foo;`
-                        let _ = module_path.pop();
-                        source = last_segment;
-                        if rename.is_none() {
-                            ident = last_segment;
-                        }
-                    }
-                } else {
-                    // Disallow `self`
-                    if source.name == keywords::SelfValue.name() {
-                        resolve_error(self,
-                                      use_tree.span,
-                                      ResolutionError::SelfImportsOnlyAllowedWithin);
-                    }
+        debug!("build_reduced_graph_for_use_tree: module_path={:?}", module_path);
 
-                    // Disallow `use $crate;`
-                    if source.name == keywords::DollarCrate.name() && path.segments.len() == 1 {
-                        let crate_root = self.resolve_crate_root(source);
-                        let crate_name = match crate_root.kind {
-                            ModuleKind::Def(_, name) => name,
-                            ModuleKind::Block(..) => unreachable!(),
-                        };
-                        source.name = crate_name;
-                        if rename.is_none() {
-                            ident.name = crate_name;
-                        }
+        let include_relative_import =
+            self.session.rust_2018() &&
+            self.session.features_untracked().relative_imports &&
+            if let Some(second_segment) = module_path.get(1) {
+                module_path[0].name == keywords::CrateRoot.name() &&
+                self.extern_prelude.contains(&second_segment.name)
```

Comment:
Talking to @joshtriplett, turns out this condition is wrong, we should *always* fork paths that don't start with `::` or a keyword.

Reply:
@petrochenkov I guess with `relative_imports`, we'd have to separate relative `use` paths (which would be the default?) from allowing to resolve through an `extern` crate (which would be gated by `extern_prelude`), but preferably still detect ambiguities when a crate *could* be loaded?

Reply:
Do you mean something like this?
```
mod rustc {}

// refers to the local module rustc as a relative path
// however, there's an extern crate `rustc` that can be loaded even if it's not in extern prelude
use rustc;
```

We can certainly detect this "ambiguity" in a post-processing step if `rustc` was successfully resolved locally, and possibly report an error.